app/models/mail_server_log.rb

```diff
@@ -218,7 +218,19 @@ def line(opts = {})
   end
 
   def delivery_status
-    @delivery_status ||= line(:decorate => true).delivery_status
+    # read in the status from the database if it's available
+    if attributes['delivery_status'].present?
+      DeliveryStatusSerializer.load(read_attribute(:delivery_status))
+    else
+      # attempt to parse the status from the log line and store if successful
+      decorated = line(:decorate => true)
+      if decorated && decorated.delivery_status
+        self.delivery_status = DeliveryStatusSerializer
+                                 .load(decorated.delivery_status)
+        save
```

Comment:
This method is fine, but I don't think we should call save here. Its really confusing that a record could be saved by calling `#delivery_status`. Instead, I think its best to initialize the delivery status just before we save for the first time – that way the serializer should handle writing the value and subsequent loads/saves should just get the record from the DB?

```ruby
before_create do
  delivery_status
end
```